core/chains/evm/logpoller/orm.go

```diff
@@ -412,3 +412,45 @@ func validateTopicIndex(index int) error {
 	}
 	return nil
 }
+
+// SelectIndexedLogsWithSigsExcluding query's for logs that have signature A and exclude logs that have a corresponding signature B, matching is done based on the topic index
+func (o *ORM) SelectIndexedLogsWithSigsExcluding(sigA, sigB common.Hash, topicIndex int, address common.Address, startBlock, endBlock int64, confs int, qopts ...pg.QOpt) ([]Log, error) {
+	if err := validateTopicIndex(topicIndex); err != nil {
+		return nil, err
+	}
+
+	q := o.q.WithOpts(qopts...)
+	var logs []Log
+
+	err := q.Select(&logs, `
+		SELECT *
+		FROM   evm_logs
+		WHERE  evm_chain_id = $1
+		AND    address = $2
+		AND    event_sig = $3
+		AND block_number BETWEEN $6 AND $7
+		AND (block_number + $8) <= (SELECT COALESCE(block_number, 0) FROM evm_log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)
+		
+		EXCEPT
+		
+		SELECT     a.*
+		FROM       evm_logs AS a
+		INNER JOIN evm_logs B
+		ON         a.evm_chain_id = b.evm_chain_id
+		AND        a.address = b.address
+		AND        a.topics[$5] = b.topics[$5]
+		AND        a.event_sig = $3
+		AND        b.event_sig = $4
+		AND (b.block_number + $8) <= (SELECT COALESCE(block_number, 0) FROM evm_log_poller_blocks WHERE evm_chain_id = $1 ORDER BY block_number DESC LIMIT 1)
```

Comment:
hmm so the block range only applies to the first event but the confs apply to both? Wouldn't we want the range to apply to both events as well?

Reply:
Seems like we should at least add`block_number >= $6`, which might speed up the query if there's a long history stored.

It's not obvious to me why we'd want to add the upper limit though.  If there's an upper limit for the block range, that seems like it would apply to the request you're looking for.  If it's already been fulfilled after the top of the range you probably don't want it... because you're looking for only unfulfilled requests, right?

Reply:
I guess there could be more general use cases... probably best to add some clear documentation on how this works, whatever we decide.

Reply:
Change in [`e2f04b0` (#8870)](https://github.com/smartcontractkit/chainlink/pull/8870/commits/e2f04b04bf86327a1c1598d65d2cf4d33cfd0629) so that the block range now applies to both. 
I don't have enough context on the use cases to have a strong opinion on the upper limit but I slightly favor removing it...

Reply:
I don't have a strong opinion on whether the upper limit should be there, my point was that if it is, it should apply only to event A, not to both A and B.  So I see the code as having been functionally correct before this change, but incorrect after it.  If we decide it should apply to both of them, we need to document that as it may not be obvious to the user.

Whether the lower limit applies to both events or only to A doesn't affect the result, but having it apply to both can improve performance, so I agree with that part of the change.

Reply:
Oh, I guess it could affect the result in general, unless we add a clause that ensures the timestamp for event A is earlier than the timestamp for event B.  I'm not sure what the use case would be outside of the request / fulfillment example... maybe I'm thinking too narrowly about the application.

Reply:
yeah I also am unsure what future use cases would be, but it does seem possible that a client may want to examine a specific historical range